README
PDL-Stats

This is a collection of statistics modules in Perl Data Language, with a quick-start guide for non-PDL people.

They make perldl--the simple shell for PDL--work like a teenie weenie R, but with PDL threading--"the fast (and automagic) vectorised iteration of 'elementary operations' over arbitrary slices of multidimensional data"--on procedures including t-test, ordinary least squares regression, and kmeans.

Of course, they also work in perl scripts.

DEPENDENCIES

PDL (>= 2.4.4)
    Perl Data Language. Preferably installed with a Fortran compiler. A few methods (logistic regression and all plotting methods) will only work with a Fortran compiler and some methods (ordinary least squares regression and pca) work much faster with a Fortran compiler.

GSL
    GNU Scientific Library. This is required by PDL::Stats::Distr and PDL::GSL::CDF, the latter of which provides p-values for PDL::Stats::GLM. GSL is otherwise NOT required for the core PDL::Stats modules to work, ie Basic, Kmeans, and GLM.

INSTALLATION

To install all included modules, run the following commands in the current dir:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for the modules with the
perldoc command.

    perldoc PDL::Stats
    perldoc PDL::Stats::Basic
    perldoc PDL::Stats::Distr
    perldoc PDL::Stats::GLM
    perldoc PDL::Stats::Kmeans
    perldoc PDL::Stats::TS
    perldoc PDL::GSL::CDF

You can also look for information at:

    Search CPAN
      http://search.cpan.org/dist/PDL-Stats/

    Sourceforge
      http://pdl-stats.sourceforge.net

    Mailing list
      https://lists.sourceforge.net/lists/listinfo/pdl-stats-help


COPYRIGHT AND LICENCE
~~~~~~~~~~~~ ~~~~~ ~~~~~~~~ ~~~~~ ~~~ `` ><(((">

Copyright (C) 2009 Maggie J. Xiong  <maggiexyz users.sourceforge.net>

All rights reserved. There is no warranty. You are allowed to redistribute this software / documentation as described in the file COPYING in the PDL distribution.


Changes
Revision history for PDL-Stats

0.5.3   04/18/2011
        stats_glm.t pca() test SLATEC vs. MatrixOps precision and sign fix.

0.5.2   04/17/2011
        stats_glm.t pca() test precision fix.

0.5.1   04/16/2011
        ***GLM pca() return value names updated*** to be more compatible with the literature.
        GLM added plot_scores(); plot_scree() name changed to plot_screes() for consistency across plotting functions.
        Updated plotting functions to default to png in Windows because GW doesn't reopen after close.

0.5.0	04/02/2011
		Updated tests to use __DATA__ instead of separate test data files.
		GLM added interaction_code(); bug fix for dummy_code().
		GLM added ols_rptd()!
        GLM added plot_residual().
		GLM eval wrapped optional modules; replaced perl reverse sort with slice (-1:0).
        Basic added binomial_test().
        TS updated plot_season() to season_m(); deprecated a few old function names, filt_exp(), filt_ma(), and dsea().

0.4.3	04/08/2010
		Distr::plot_distr() bug fix.
		Kmeans::kmeans() pod update.

0.4.2	03/27/2010
		rtable() added to PDL::Basic. get_data() deprecated. Preparing for transitioning to using rcols() for IO.
		perldl online help support! (ie "? stdv" in perldl works)

0.4.1	02/02/2010
		GLM anova anova_rptd IVNM default val matched to [].
		Makefile.PL exit 0 without PDL::Lite.
		TS added wmape
		pod fix for CPAN

0.4.0	01/24/2010
		GLM anova_rptd supports between subject factors! We are now beta!

0.3.2	01/08/2010
		Kmeans kmeans CNTRD checks undef besides PDL->null. opt{V} printout update
		Kmeans added which_cluster
		Kmeans pca_cluster does scree plot with cutoff line
		GLM logistic pod update
		GLM anova anova_rptd pca plot_scree plotting options update

0.3.1	12/16/2009
		Removed pms--relying on pdl-stats.sourceforge.net for web documentation instead of CPAN. Minor pod update.
		GLM::ols_t removed $y->squeeze, so that edge case like ols_t on 1 obs threads properly (y[1,10]->ols_t(x[1,10]->dummy(1), {const=>0}) gives b[1,10]).
		GLM::anova_rptd se fix.
		GLM::anova and anova_rptd plot se fix.
	    PDL::Stats::TS included. ***alpha alpha*** code! Included because I'm tired of maintaining it separately from PDL::Stats.

0.3.0	12/03/2009
        GLM::anova_rptd! works for purely within designs
        GLM::anova bug fix--not messing up caller @ if passed @ ref in IV instead of pdl
	    Change prereq from PDL to PDL::Lite
	    Stats.pm quick-start guide update
	    t/stats_glm logistic test modified to use more proper numbers in x

0.2.8	11/04/2009
	    Fixed minor bug (mismatched nseed and self cluster pdls. oops) in switching to use cluster assignment as stop criteria. btw initial R2 now based on self instead of seeds.

0.2.7	11/04/2009
        GLM::plot_scree allows directly passing NCOMP. plots cutoff line for screes if requested
        kmeans stops when cluster assignment no longer changes instead of using R2CRT.
        If not specified, Kmeans::pca_cluster studies scree plot to determine NCOMP
	    Basic::get_data STDERR message fix: data pdl o x v, NOT v x o

0.2.6	10/12/2009
        anova bad value se bug fix
        centroid bad value empty cluser bug fix
        basic, glm, distr rid of nan and use bad value instead

0.2.5   09/17/2009
        PREREQ_FATAL => 1 in Makefile.PL
        added pca_cluster in Kmeans.
        anova one-way bug fix.
        plot_means one-way bug fix.

0.2.4   09/02/2009
        added corr_table in PDL::Stats::Basic.
        PDL::Stats::GLM::pca uses corr_table.
        kmeans no longer quits with empty cluster. kmeans 3d threading bug fix.
        use short-circuit ISBAD(a) || ISBAD(b) instead of ISGOOD(a) && ISGOOD(b) in bad value processing.

0.2.3   08/24/2009
        ***anova interface change*** IV names now passed in \%opt instead of @ ref. I appologize for the interface change, but it matches the new method plot_means, and makes it easier to pass IVs themselves. I promise that such interface changes will be extremely rare, if at all.
        anova now supports bad value in DV.
        Added plot_means in PDL::Stats::GLM. anova plots highest order interaction by default.
        Added plot_scree in PDL::Stats::GLM. pca does scree plot by default.
        Added pca_sorti in PDL::Stats::GLM.
        plot_distr in PDL::Stats::Distr accepts WIN option
        Set prerequisite PDL => 2.4.4, which has bad value support by default

0.2.2   07/23/2009
        stats_distr.t moved form t/ to Distr/t, ie skip the test if no GSL.
        more kmeans fix (4d, seeding). added seeded kmeans tests.

0.2.1   07/22/2009
        kmeans 4d fix. added pms dir for cpan pod parse

0.2.0   07/21/2009
        kmeans threads! ss/ms bug fix. more tests on kmeans added.

0.1.3   07/17/2009
        quick-start for non-PDL people in Stats.pm pod
        ols_t uses double internally. GLM pod update
        r2_change threads (switched to use ols_t instead of ols)
        get_data which_id moved to Basic from GLM
        tests grouped in PDL-Stats/t except for PDL::GSL::CDF

0.1.2   07/15/2009
        Cleaned up some versioning stuff in Stats.pm
        GLM::get_data TYPE default to double. MISSN handling update; set to BAD by default.
        Case insensitive option passing.

0.1.1   07/14/2009
        Fixed PDL::Stats::GLM::ols b_p value for b_t < 0.

0.1.0   07/13/2009
        First version, released on an unsuspecting world.


pms/Kmeans.pm

#
# GENERATED WITH PDL::PP! Don't modify!
#
package PDL::Stats::Kmeans;

@EXPORT_OK  = qw(  random_cluster iv_cluster PDL::PP _random_cluster PDL::PP assign PDL::PP centroid PDL::PP _d_p2l );
%EXPORT_TAGS = (Func=>[@EXPORT_OK]);

use PDL::Core;
use PDL::Exporter;
use DynaLoader;



   
   @ISA    = ( 'PDL::Exporter','DynaLoader' );
   push @PDL::Core::PP, __PACKAGE__;
   bootstrap PDL::Stats::Kmeans ;





use Carp;
use PDL::LiteF;
use PDL::NiceSlice;
use PDL::Stats::Basic;

=head1 NAME

PDL::Stats::Kmeans -- classic k-means cluster analysis 

=head1 DESCRIPTION

Assumes that we have data pdl dim [observation, variable] and the goal is to put observations into clusters based on their values on the variables. The terms "observation" and "variable" are quite arbitrary but serve as a reminder for "that which is being clustered" and "that which is used to cluster".

The terms FUNCTIONS and METHODS are arbitrarily used to refer to methods that are threadable and methods that are non-threadable, respectively.

=head1 SYNOPSIS

    use PDL::LiteF;
    use PDL::NiceSlice;
    use PDL::Stats;

    my ($data, $idv, $ido) = get_data( $file );

    my ($cluster, $centroid, $ss_this, $ss_last, $ss_centroid);

      # start out with 8 random clusters
    $cluster = random_cluster( $data->dim(0), 8 );

    ($centroid, $ss_centroid) = $data->centroid( $cluster );
    $ss_this = $ss_centroid->sum;

      # iterate to minimize total ss
      # stop when change in ss is less than $crit amount 
    do {
      $ss_last = $ss_this;
      $cluster = $data->assign( $centroid );

      ($centroid, $ss_centroid) = $data->centroid( $cluster );
      $ss_this = $ss_centroid->sum;
    }
    while ( $ss_last - $ss_this > $crit );

or just do

    my %result = $data->kmeans( \%opt );
    print "$_\t$result{$_}\n" for (sort keys %result);

plot the clusters if there are only 2 vars

    use PDL::Graphics::PGPLOT::Window;

    my ($win, $c);
    $win = pgwin(Dev=>'/xs');
    $win->env($data( ,0)->minmax, $data( ,1)->minmax);

    $win->points( $data->dice_axis(0,which($m{cluster}->(,$_)))->dog,
                  {COLOR=>++$c} )
      for (0..$m{cluster}->dim(1)-1);

=cut







=head1 FUNCTIONS



=cut





=head2 random_cluster

=for sig

  Signature: (byte [o]cluster(o,c); int obs=>o; int clu=>c)

=for ref

Creates masks for random mutually exclusive clusters. Accepts two parameters, num_obs and num_cluster. Extra parameter turns into extra dim in mask. May loop a long time if num_cluster approaches num_obs because empty cluster is not allowed.

=for usage

    my $cluster = random_cluster( $num_obs, $num_cluster );

=cut

  # can't be called on pdl
sub random_cluster {
  my ($obs, $clu) = @_;
    # extra param in @_ made into extra dim
  my $cluster = zeroes @_;
  do {
    $cluster->inplace->_random_cluster();
  } while (PDL::any $cluster->sumover == 0 );
  return $cluster;
}





*_random_cluster = \&PDL::_random_cluster;




=head2 assign

=for sig

  Signature: (data(o,v); centroid(c,v); byte [o]cluster(o,c))



=for ref

Takes data pdl dim [obs x var] and centroid pdl dim [cluster x var] and returns mask dim [obs x cluster] to cluster membership. An obs is assigned to the first cluster with the smallest distance (ie sum squared error) to cluster centroid. With bad value, obs is assigned by smallest mean squared error across variables.

=for usage

    perldl> $centroid = ones 2, 3
    perldl> $centroid(0,) .= 0
    perldl> p $centroid
    [
     [0 1]
     [0 1]
     [0 1]
    ]

    perldl> $b = qsort( random 4, 3 )
    perldl> p $b
    [
     [0.022774068 0.032513883  0.13890034  0.30942479]
     [ 0.16943853  0.50262636  0.56251531   0.7152271]
     [ 0.23964483  0.59932745  0.60967495  0.78452117]
    ]
      # notice that 1st 3 obs in $b are on average closer to 0 
      # and last obs closer to 1
    perldl> p $b->assign( $centroid )
    [
     [1 1 1 0]    # cluster 0 membership
     [0 0 0 1]    # cluster 1 membership
    ]
  

=for bad

assign does handle bad values.
It will set the bad-value flag of all output piddles if the flag is set for any of the input piddles.


=cut






*assign = \&PDL::assign;




=head2 centroid

=for sig

  Signature: (data(o,v); cluster(o,c); float+ [o]m(c,v); float+ [o]ss(c,v))


=for ref

Takes data dim [obs x var] and mask dim [obs x cluster], returns mean and ss (ms when data contains bad values) dim [cluster x var], using data where mask = 1. Multiple cluster membership for an obs is okay. If a cluster is empty all means and ss are set to zero for that cluster.

=for usage

      # data is 10 obs x 3 var
    perldl> p $d = sequence 10, 3
    [
     [ 0  1  2  3  4  5  6  7  8  9]
     [10 11 12 13 14 15 16 17 18 19]
     [20 21 22 23 24 25 26 27 28 29]
    ]
      # create two clusters by value on 1st var
    perldl> p $a = $d( ,(0)) <= 5
    [1 1 1 1 1 1 0 0 0 0]
 
    perldl> p $b = $d( ,(0)) > 5
    [0 0 0 0 0 0 1 1 1 1]

    perldl> p $c = cat $a, $b
    [
     [1 1 1 1 1 1 0 0 0 0]
     [0 0 0 0 0 0 1 1 1 1]
    ]

    perldl> p $d->centroid($c)
      # mean for 2 cluster x 3 var
    [
     [ 2.5  7.5]
     [12.5 17.5]
     [22.5 27.5]
    ]
      # ss for 2 cluster x 3 var
    [
     [17.5    5]
     [17.5    5]
     [17.5    5]
    ]

  

=for bad

centroid does handle bad values.
It will set the bad-value flag of all output piddles if the flag is set for any of the input piddles.


=cut






*centroid = \&PDL::centroid;



sub _scree_ind {
  # it's a heuristic--whether we can get "good" results depends on
  # the number of components in $self.

  my ($self) = @_;

  $self = $self->squeeze;
  $self->ndims > 1 and
    croak "1D pdl only please";

  my $a = zeroes 2, $self->nelem;
  $a((0), ) .= sequence $self->nelem;
  $a((1), ) .= $self;

  my $d = _d_point2line( $a, $a( ,(0)), $a( ,(-1)) );

  return $d->maximum_ind;
}

sub _d_point2line {
  my ($self, $p1, $p2) = @_;

  for ($self, $p1, $p2) {
    $_->dim(0) != 2 and
      carp "point pdl dim(0) != 2";
  }

  return _d_p2l( $self->mv(0,-1)->dog, $p1->mv(0,-1)->dog, $p2->mv(0,-1)->dog );
}





*_d_p2l = \&PDL::_d_p2l;



=head2 kmeans

=for ref

Implements classic kmeans cluster analysis. Tries several different random-seeding and clustering in parallel. Stops when cluster assignment no longer changes. Returns the best result in terms of R2.

Alternatively, if a centroid is provided, clustering will proceed from the centroid and there is no random-seeding or multiple tries.

kmeans supports bad value*.

=for options

Default options (case insensitive):

  V       => 1,          # prints simple status
  FULL    => 0,          # returns results for all seeding rounds

  CNTRD   => PDL->null,  # clu x var. optional. disables next 3 opts

  NTRY    => 5,          # num of seeding rounds
  NSEED   => 1000,       # num of initial seeds, use NSEED up to max obs
  NCLUS   => 8,          # num of clusters

=for usage

Usage:

  # suppose we have 4 person's ratings on 5 movies

  perldl> p $rating = ceil( random(4, 5) * 5 ) 
  [
   [3 2 2 3]
   [2 4 5 4]
   [5 3 2 3]
   [3 3 1 5]
   [4 3 3 2]
  ]

  # we want to put the 4 persons into 2 groups

  perldl> %k = $rating->kmeans( {NCLUS=>2} )

  # by default prints back options used
  # as well as info for all tries and iterations

  CNTRD	=> Null
  FULL	=> 0
  NCLUS	=> 2
  NSEED	=> 4
  NTRY	=> 5
  V     => 1
  ss total:	20.5
  iter 0 R2 [0.024390244 0.024390244 0.26829268  0.4796748  0.4796748]
  iter 1 R2 [0.46341463 0.46341463  0.4796748  0.4796748  0.4796748]

  perldl> p "$_\t$k{$_}\n" for (sort keys %k)

  R2      0.479674796747968
  centroid       # mean ratings for 2 group x 5 movies
  [
   [         3  2.3333333]
   [         2  4.3333333]
   [         5  2.6666667]
   [         3          3]
   [         4  2.6666667]
  ]
 
  cluster        # 4 persons' membership in two groups
  [
   [1 0 0 0]
   [0 1 1 1]
  ]
  
  n       [1 3]  # cluster size
  ss
  [
   [         0 0.66666667]
   [         0 0.66666667]
   [         0 0.66666667]
   [         0          8]
   [         0 0.66666667]
  ]

Now, for the valiant, kmeans is threadable. Say you gathered 10 persons' ratings on 5 movies from 2 countries, so the data is dim [10,5,2], and you want to put the 10 persons from each country into 3 clusters, just specify NCLUS => [3,1], and there you have it. The key is for NCLUS to include $data->ndims - 1 numbers. The 1 in [3,1] turns into a dummy dim, so the 3-cluster operation is repeated on both countries. Similarly, when seeding, CNTRD needs to have ndims that at least match the data ndims. Extra dims in CNTRD will lead to threading (convenient if you want to try out different centroid locations, for example, but you will have to hand pick the best result). See stats_kmeans.t for examples w 3D and 4D data.

*With bad value, R2 is based on average of variances instead of sum squared error.

=cut

*kmeans = \&PDL::kmeans;
sub PDL::kmeans {
  my ($self, $opt) = @_;
  my %opt = (
    V          => 1,
    FULL       => 0,

    CNTRD      => PDL->null,
      # next 3 opt not used if centroid is provided
    NTRY       => 5,
    NSEED      => 1000,
    NCLUS      => 3,
  );
  $opt and $opt{uc $_} = $opt->{$_} for (keys %$opt);
  if ($opt{CNTRD}->nelem) {
    $opt{NTRY}  = 1;
    $opt{NSEED} = $self->dim(0);
  }
  else {
    $opt{NSEED} = pdl($self->dim(0), $opt{NSEED})->min;
  }
  $opt{V} and print STDERR "$_\t=> $opt{$_}\n" for (sort keys %opt);
 
  my $ss_ms = $self->badflag?  'ms' : 'ss';
  my $ss_total
    = $self->badflag?  $self->var->average : $self->ss->sumover;
  $opt{V} and print STDERR "overall $ss_ms:\t$ss_total\n";

  my ($centroid, $ss_cv, $R2, $clus_this, $clus_last);

    # NTRY made into extra dim in $cluster for threading
  my @nclus = (ref $opt{NCLUS} eq 'ARRAY')? @{$opt{NCLUS}} : ($opt{NCLUS});
  $clus_this
    = $opt{CNTRD}->nelem ?
      $self->assign( $opt{CNTRD}->dummy(-1) )  # put dummy(-1) to match NTRY
    : random_cluster($opt{NSEED}, @nclus, $opt{NTRY} )
    ;

  ($centroid, $ss_cv) = $self(0:$opt{NSEED} - 1, )->centroid( $clus_this );
    # now obs in $clus_this matches $self
  $clus_this = $self->assign( $centroid );
  ($centroid, $ss_cv) = $self->centroid( $clus_this );

  my $iter = 0;
  do {
    $R2 = $self->badflag? 1 - $ss_cv->average->average / $ss_total
        :                 1 - $ss_cv->sumover->sumover / $ss_total
        ;
    $opt{V} and print STDERR join(' ',('iter', $iter++, 'R2', $R2)) . "\n";

    $clus_last = $clus_this;
   
    $clus_this = $self->assign( $centroid );
    ($centroid, $ss_cv) = $self->centroid( $clus_this );
  }
  while ( any long(abs($clus_this - $clus_last))->sumover->sumover > 0 );

  $opt{FULL} and 
    return (
      centroid => PDL::squeeze( $centroid ),
      cluster  => PDL::squeeze( $clus_this ),
      n        => PDL::squeeze( $clus_this )->sumover,
      R2       => PDL::squeeze( $R2 ), 
      $ss_ms   => PDL::squeeze( $ss_cv ),
    );

    # xchg/mv(-1,0) leaves it as was if single dim--unlike transpose
  my $i_best = $R2->mv(-1,0)->maximum_ind;

  $R2->getndims == 1 and
    return (
      centroid => $centroid->dice_axis(-1,$i_best)->sever->squeeze,
      cluster  => $clus_this->dice_axis(-1,$i_best)->sever->squeeze,
      n        => $clus_this->dice_axis(-1,$i_best)->sever->squeeze->sumover,
      R2       => $R2->dice_axis(-1,$i_best)->sever->squeeze, 
      $ss_ms   => $ss_cv->dice_axis(-1,$i_best)->sever->squeeze,
    );

  # now for threading beyond 2D data

  # can't believe i'm using a perl loop :P

  $i_best = $i_best->flat->sever;
  my @i_best = map { $opt{NTRY} * $_ + $i_best(($_)) }
               0 .. $i_best->nelem - 1;

  my @shapes;
  for ($centroid, $clus_this, $R2) {
    my @dims = $_->dims;
    pop @dims;
    push @shapes, \@dims;
  }

  $clus_this = $clus_this->mv(-1,2)->clump(2..$clus_this->ndims-1)->dice_axis(2,\@i_best)->reshape( @{ $shapes[1] } )->sever,

  return (
    centroid =>
$centroid->mv(-1,2)->clump(2..$centroid->ndims-1)->dice_axis(2,\@i_best)->reshape( @{ $shapes[0] } )->sever,

    cluster  => $clus_this,
    n        => $clus_this->sumover,

    R2       =>
$R2->mv(-1,0)->clump(0..$R2->ndims-1)->dice_axis(0,\@i_best)->reshape( @{ $shapes[2] } )->sever,

    $ss_ms   => 
$ss_cv->mv(-1,2)->clump(2..$ss_cv->ndims-1)->dice_axis(2,\@i_best)->reshape( @{ $shapes[0] } )->sever,
  ); 
}

=head1 METHODS

=head2 iv_cluster

=for ref

Turns an independent variable into a cluster pdl. Returns cluster pdl and level-to-pdl_index mapping in list context and cluster pdl only in scalar context.

This is the method used for mean and var in anova. The difference between iv_cluster and dummy_code is that iv_cluster returns pdl dim (obs x level) whereas dummy_code returns pdl dim (obs x (level - 1)).

=for usage

Usage:

    perldl> @bake = qw( y y y n n n )
   
    # accepts @ ref or 1d pdl

    perldl> p $bake = iv_cluster( \@bake )
    [
     [1 1 1 0 0 0]
     [0 0 0 1 1 1]
    ]
    
    perldl> p $rating = sequence 6
    [0 1 2 3 4 5]

    perldl> p $rating->centroid( $bake )
    # mean for each iv level
    [
     [1 4]
    ]
    # ss
    [
     [2 2]
    ]

=cut

*iv_cluster = \&PDL::iv_cluster;
sub PDL::iv_cluster {
  my ($var_ref) = @_;

    # pdl->uniq puts elem in order. so instead list it to maintain old order
  if (ref $var_ref eq 'PDL') {
    $var_ref = $var_ref->squeeze;
    $var_ref->getndims > 1 and
      croak "multidim pdl passed for single var!";
    $var_ref = [ list $var_ref ];
  }

  my ($var, $map_ref) = _array_to_pdl( $var_ref );
  my $var_a = zeroes byte, $var->nelem, $var->max + 1;

  for my $l (0 .. $var->max) {
    my $v = $var_a( ,$l);
    $v->index( which $var == $l ) .= 1;
  }

  return wantarray? ($var_a, $map_ref) : $var_a;
}

=head2 pca_cluster

Assgin variables to components ie clusters based on pca loadings or scores. One way to seed kmeans (see Ding & He, 2004, and Su & Dy, 2004 for other ways of using pca with kmeans). Variables are assigned to their most associated component. Note that some components may not have any variable that is most associated with them, so the returned number of clusters may be smaller than NCOMP.

Default options (case insensitive):

  V     => 1,
  ABS   => 1,     # high pos and neg loadings on a comp in same cluster
  NCOMP => undef, # max number of components to consider. determined by
                  # scree plot black magic if not specified

Usage:

    # say we need to cluster a group of documents
  ($data, $idd, $idw) = get_data 'doc_word_info.txt';

  perldl> %p = $data->pca;
  perldl> $cluster  = $p{loading}->pca_cluster;

    # pca clusters var while kmeans clusters obs. hence transpose
  perldl> ($m, $ss) = $data->transpose->centroid( $cluster );
  perldl> %k = $data->transpose->kmeans( { cntrd=>$m } );

    # take a look at cluster 0 doc ids
  perldl> p join("\n", @$idd[ list which $k{cluster}->( ,0) ]);

=cut

*pca_cluster = \&PDL::pca_cluster;
sub PDL::pca_cluster {
  my ($self, $opt) = @_;

  my %opt = (
    V     => 1,
    ABS   => 1,     # high pos and neg loadings on a comp in same cluster
    NCOMP => undef, # max number of components to consider. determined by
                    # scree plot black magic if not specified
  );
  $opt and $opt{uc $_} = $opt->{$_} for (keys %$opt);

  if (!$opt{NCOMP}) {
      # here's the black magic part
    my $comps = ($self->dim(1) > 300)? int($self->dim(1) * .1)
              :                        pdl($self->dim(1), 30)->min
              ;
    my $var   = sumover($self( ,0:$comps-1) ** 2) / $self->dim(0);
    $opt{NCOMP} = _scree_ind( $var );
  }

  my $c = $self->( ,0:$opt{NCOMP}-1)->transpose->abs->maximum_ind;

  if ($opt{ABS}) {
    $c = $c->iv_cluster;
  }
  else {
    my @c = map { ($self->($_,$c($_)) >= 0)? $c($_)*2 : $c($_)*2 + 1 }
                ( 0 .. $c->dim(0)-1 );
    $c = iv_cluster( \@c );
  }
  $opt{V} and print STDERR "cluster membership mask as " . $c->info . "\n";
  return $c;
}

sub _array_to_pdl {
  my ($var_ref) = @_;

  my (%level, $l);
  $l = 0;
  for (@$var_ref) {
    !exists $level{$_} and $level{$_} = $l ++;
  } 
  return wantarray? (pdl( map { $level{$_} } @$var_ref ), \%level)
        :            pdl( map { $level{$_} } @$var_ref )
        ;
}

=head1 	REFERENCES

Ding, C., & He, X. (2004). K-means clustering via principal component analysis. Proceedings of the 21st International Conference on Machine Learning, 69, 29.

Su, T., & Dy, J. (2004). A deterministic method for initializing K-means clustering. 16th IEEE International Conference on Tools with Artificial Intelligence, 784-786.

Romesburg, H.C. (1984). Cluster Analysis for Researchers. NC: Lulu Press.

Wikipedia (retrieved June, 2009). K-means clustering. http://en.wikipedia.org/wiki/K-means_algorithm

=head1 AUTHOR

Copyright (C) 2009 Maggie J. Xiong <maggiexyz users.sourceforge.net>

All rights reserved. There is no warranty. You are allowed to redistribute this software / documentation as described in the file COPYING in the PDL distribution.

=cut



;



# Exit with OK status

1;

		   